internal/gadget-runner/gadgetrunner.go

```diff
@@ -0,0 +1,125 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package gadgetrunner handles initializing gadgets and installed operators before
+handing them over to a specified runtime.
+*/
+package gadgetrunner
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/parser"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/runtime"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+// GadgetRunner handles running gadgets by the gadget interface; it orchestrates the whole lifecycle of the gadget
+// instance and communicates with gadget and runtime.
+type GadgetRunner struct {
+	ctx          context.Context
+	id           string
+	gadget       gadgets.Gadget
+	gadgetParams *params.Params
+	runtime      runtime.Runtime
+	parser       parser.Parser
+	operators    operators.Operators
+	logger       logger.Logger
+	result       []byte
+	resultError  error
+}
+
+func NewGadgetRunner(
+	ctx context.Context,
+	id string,
+	runtime runtime.Runtime,
+	gadget gadgets.Gadget,
+	parser parser.Parser,
+	logger logger.Logger,
+) *GadgetRunner {
+	return &GadgetRunner{
+		ctx:     ctx,
+		id:      id,
+		gadget:  gadget,
+		runtime: runtime,
+		parser:  parser,
+		logger:  logger,
+	}
+}
+
+func (r *GadgetRunner) ID() string {
+	return r.id
+}
+
+func (r *GadgetRunner) Context() context.Context {
+	return r.ctx
+}
+
+func (r *GadgetRunner) Parser() parser.Parser {
+	return r.parser
+}
+
+func (r *GadgetRunner) Runtime() runtime.Runtime {
+	return r.runtime
+}
+
+func (r *GadgetRunner) Gadget() gadgets.Gadget {
+	return r.gadget
+}
+
+func (r *GadgetRunner) Operators() operators.Operators {
+	return r.operators
+}
+
+func (r *GadgetRunner) Logger() logger.Logger {
+	return r.logger
+}
+
+func (r *GadgetRunner) SetResult(result []byte, resultError error) {
+	r.result = result
+	r.resultError = resultError
+}
+
+func (r *GadgetRunner) GetResult() ([]byte, error) {
+	return r.result, r.resultError
+}
+
+func (r *GadgetRunner) GadgetParams() *params.Params {
+	return r.gadgetParams
+}
+
+// RunGadget is the main function of GadgetRunner and controls the lifecycle of the gadget
+func (r *GadgetRunner) RunGadget(
+	runtimeParams *params.Params,
+	operatorParamCollection params.Collection,
+	operatorPerGadgetParamCollection params.Collection,
+	gadgetParams *params.Params,
+) error {
+	r.gadgetParams = gadgetParams
+	r.operators = operators.GetOperatorsForGadget(r.gadget)
+	err := r.operators.Init(operatorParamCollection)
```

Comment:
Why to init the operators just before running the gadget? Isn't it something that should happen sooner on the application lifecycle?

Reply:
Also, when/where to call `operator.Close()`?

Reply:
For now that should happen before exiting. Later on we could have them stopped after some period to free up some resources and started again on-demand.

Reply:
I think it should be handled globally by the framework (maybe even the operator-registry), there shouldn't be anything done in each operator. Right now the only operator is the container-collection that will probably also always be the most heavy weight (even more heavy when used in combination with the k8s API server in a cluster environment) - and right now that is running all the time, getting background updates for containers and storing them.

As we only use that information when gadgets are running anyways, the optimization would be to not run the operator from start, but initialize it on-demand when the first gadget uses it. I wouldn't directly stop it after the gadget run, though, as the spin-up might be costly and we can assume that debugging is ongoing when launching a gadget. Instead, I'd only close the operators after some time has passed (10 minutes or so for example).

In a big cluster that could save a lot of API calls and wasted CPU cycles.

Reply:
I agree on this. We can explore later on the different possibilities about Init(), Close(), but I think the current model we have is flexible enough.